indexbs/indexbacked_bs.go

```diff
@@ -0,0 +1,194 @@
+	// instantiate the blockstore cache
+	bslru, err := lru.NewWithEvict(maxCacheSize, func(_ interface{}, val interface{}) {
+		// ensure we close the blockstore for a shard when it's evicted from the cache so dagstore can gc it.
+		abs := val.(*accessorWithBlockstore)
+		abs.sa.Close()
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create lru cache for read only blockstores")
+	}
+
+	return &IndexBackedBlockstore{
+		d:               d,
+		shardSelectF:    shardSelector,
+		blockstoreCache: bslru,
+	}, nil
+}
+
+func (ro *IndexBackedBlockstore) Get(ctx context.Context, c cid.Cid) (b blocks.Block, finalErr error) {
+	logbs.Debugw("Get called", "cid", c)
+	defer func() {
+		if finalErr != nil {
+			logbs.Debugw("Get: got error", "cid", c, "error", finalErr)
+		}
+	}()
+
+	mhash := c.Hash()
+
+	// fetch all the shards containing the multihash
+	shards, err := ro.d.ShardsContainingMultihash(ctx, mhash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch shards containing the block: %w", err)
+	}
+	if len(shards) == 0 {
+		return nil, ErrBlockNotFound
+	}
+
+	// do we have a cached blockstore for a shard containing the required block ? If yes, serve the block from that blockstore
+	for _, sk := range shards {
+		// a valid cache hit here updates the priority of the shard's blockstore in the LRU cache.
+		val, ok := ro.blockstoreCache.Get(sk)
+		if !ok {
+			continue
+		}
+
+		rbs := val.(*accessorWithBlockstore).bs
+		blk, err := rbs.Get(ctx, c)
+		if err != nil {
+			// we know that the cid we want to lookup belongs to a shard with key `sk` and
+			// so if we fail to get the corresponding block from the blockstore for that shards, something has gone wrong
+			// and we should remove the blockstore for that shard from our cache.
+			ro.blockstoreCache.Remove(sk)
+			continue
+		}
+
+		// add the block to the block cache
+		logbs.Debugw("Get: returning from block store cache", "cid", c)
+		return blk, nil
+	}
+
+	// ---- we don't have a cached blockstore for a shard that can serve the block -> let's build one.
+
+	// select a valid shard that can serve the retrieval
+	sk, err := ro.shardSelectF(c, shards)
+	if err != nil {
+		return nil, ErrBlockNotFound
+	}
+
+	// load blockstore for the selected shard and try to serve the cid from that blockstore.
+	resch := make(chan dagstore.ShardResult, 1)
+	if err := ro.d.AcquireShard(ctx, sk, resch, dagstore.AcquireOpts{}); err != nil {
+		return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, err)
+	}
+	var res dagstore.ShardResult
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case res = <-resch:
+		if res.Error != nil {
+			return nil, fmt.Errorf("failed to acquire shard %s: %w", sk, res.Error)
+		}
+	}
+
+	sa := res.Accessor
+	bs, err := sa.Blockstore()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load read only blockstore for shard %s: %w", sk, err)
+	}
+
+	blk, err := bs.Get(ctx, c)
+	if err != nil {
```

Comment:
Let's check here if the block is not found, and if so, return `ErrBlockNotFound`

Reply:
The thing is, that should never happen because we've already checked that the shard contains the requested cid and so the bockstore for that shard should have the requested cid. Even if it is a `ErrBlockNotFound `, something has gone badly wrong here and so we shouldn't return a simple `ErrBlockNotFound ` here.

Reply:
Ok fair enough